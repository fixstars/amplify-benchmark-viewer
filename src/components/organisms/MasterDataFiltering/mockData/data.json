[
  { "name": "Tsp" },
  { "name": "Qap" },
  { "name": "MaxCut" },
  { "name": "Sudoku" },
  { "name": "Cvrp" }
]
