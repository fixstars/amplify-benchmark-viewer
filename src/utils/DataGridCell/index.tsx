/**
 * Copyright (c) Fixstars Corporation and Fixstars Amplify Corporation.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { LinkButton } from 'components/atoms/LinkButton'

const clientColors = [
  '#0f66dc',
  '#f57bae',
  '#e75e75',
  '#e5f97c',
  '#1b000d',
  '#f6a695',
  '#692981',
  '#3cde29',
  '#591414',
  '#d66594',
  '#2d884a',
  '#400763',
  '#9d19f4',
  '#fcdeb9',
  '#1f4c48',
  '#909743',
  '#2dc1c8',
  '#c17913',
  '#b2a1f6',
]

const instanceColors = [
  '#ff66ff',
  '#0f66dc',
  '#f57bae',
  '#e75e75',
  '#e5f97c',
  '#1b000d',
  '#f6a695',
  '#692981',
  '#3cde29',
  '#591414',
  '#d66594',
  '#2d884a',
  '#400763',
  '#9d19f4',
  '#fcdeb9',
  '#718de6',
  '#1f4c48',
  '#909743',
  '#2dc1c8',
  '#c17913',
  '#b2a1f6',
  '#9a916d',
  '#406a0f',
  '#004b31',
  '#18729f',
  '#9997c3',
  '#9cf0ec',
  '#ce843c',
  '#2d326e',
  '#76ed60',
  '#d7f7e8',
  '#bb15d7',
  '#87a7d8',
  '#94053a',
  '#e5c8a6',
  '#e2f2c0',
]

export interface ProblemInstances {
  readonly [key: string]: ReadonlyArray<string> | undefined
}

const renderClientsCell = (key: string, clients: ReadonlyArray<string>) => {
  const getBackgroundColor = (client: string): string => {
    const cache = JSON.parse(localStorage.getItem('clientColorCache') ?? '{}')
    let cacheIndex = JSON.parse(
      localStorage.getItem('clientColorIndexCache') ?? '-1',
    )
    let backgroundColor = cache[client]

    if (backgroundColor === undefined) {
      const length = clientColors.length
      cacheIndex = cacheIndex >= length ? 0 : cacheIndex + 1
      backgroundColor = clientColors[cacheIndex]
      cache[client] = backgroundColor
      localStorage.setItem('clientColorCache', JSON.stringify(cache))
      localStorage.setItem('clientColorIndexCache', JSON.stringify(cacheIndex))
      return backgroundColor
    }

    return backgroundColor
  }

  return clients.map((client) => (
    <div key={`${key}_${client}`} style={{ marginRight: 10 }}>
      <LinkButton
        label={client}
        link={`/clients/${client}`}
        backgroundColor={getBackgroundColor(client)}
      />
    </div>
  ))
}

const renderProblemInstancesCell = (
  key: string,
  problemInstances: ProblemInstances,
) => {
  const getBackgroundColor = (instance: string): string => {
    const cache = JSON.parse(localStorage.getItem('instanceColorCache') ?? '{}')
    let cacheIndex = JSON.parse(
      localStorage.getItem('instanceColorIndexCache') ?? '-1',
    )
    let backgroundColor = cache[instance]

    if (backgroundColor === undefined) {
      const length = instanceColors.length
      cacheIndex = cacheIndex >= length ? 0 : cacheIndex + 1
      backgroundColor = instanceColors[cacheIndex]
      cache[instance] = backgroundColor
      localStorage.setItem('instanceColorCache', JSON.stringify(cache))
      localStorage.setItem(
        'instanceColorIndexCache',
        JSON.stringify(cacheIndex),
      )
      return backgroundColor
    }

    return backgroundColor
  }
  const pairs = Object.entries(problemInstances)
  pairs.sort(function (p1, p2) {
    const p1Key = p1[0]
    const p2Key = p2[0]
    if (p1Key < p2Key) {
      return -1
    }
    if (p1Key > p2Key) {
      return 1
    }
    return 0
  })
  problemInstances = Object.fromEntries(pairs)
  return Object.keys(problemInstances).map((className) => (
    <div
      key={`${key}_${className}`}
      style={{ margin: '5px 0 5px 0', display: 'flex' }}
    >
      <span style={{ marginRight: 10 }}>{className}:</span>
      <div style={{ display: 'flex', flexWrap: 'wrap' }}>
        {problemInstances[className]?.map((instance) => (
          <div
            key={`${key}_${className}_${instance}`}
            style={{ marginRight: 10, marginBottom: 10 }}
          >
            <LinkButton
              label={instance}
              link={`/classes/${className}/instances/${instance}`}
              backgroundColor={getBackgroundColor(instance)}
            />
          </div>
        ))}
      </div>
    </div>
  ))
}

export { renderClientsCell, renderProblemInstancesCell }
