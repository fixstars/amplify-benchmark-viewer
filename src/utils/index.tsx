export * from './DataGridCell'
export * from './statistics'
