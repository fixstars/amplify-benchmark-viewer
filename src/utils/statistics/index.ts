function quartile(data: readonly number[], q: number) {
  const sortedData = data.slice().sort((a, b) => a - b)
  const pos = Math.floor((sortedData.length - 1) * q)
  return sortedData[pos]
}
function quartile0(data: readonly number[]) {
  return quartile(data, 0)
}
function quartile25(data: readonly number[]) {
  return quartile(data, 0.25)
}

function quartile50(data: readonly number[]) {
  return quartile(data, 0.5)
}

function quartile75(data: readonly number[]) {
  return quartile(data, 0.75)
}

function quartile100(data: readonly number[]) {
  return quartile(data, 1)
}

export { quartile0, quartile25, quartile50, quartile75, quartile100 }
