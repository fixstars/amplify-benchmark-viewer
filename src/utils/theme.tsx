/**
 * Copyright (c) Fixstars Corporation and Fixstars Amplify Corporation.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { createTheme } from '@mui/material/styles'

export const theme = createTheme({
  palette: {
    primary: {
      main: '#111',
    },
  },
  typography: {
    h2: {
      fontSize: '1.5rem',
    },
  },
  components: {
    MuiTooltip: {
      styleOverrides: {
        tooltip: {
          backgroundColor: '#111',
          color: '#f4f1f4',
          font: '10px',
        },
      },
    },
  },
})
