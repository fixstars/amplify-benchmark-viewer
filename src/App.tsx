/**
 * Copyright (c) Fixstars Corporation and Fixstars Amplify Corporation.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { ThemeProvider } from '@mui/material'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import {
  ClientListPage,
  ClientPage,
  LabelListPage,
  ProblemListPage,
  ProblemPage,
  UploadDataPage,
} from 'pages'
import { Route, Routes } from 'react-router-dom'
import { RecoilRoot } from 'recoil'
import { theme } from 'utils/theme'

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
})

function App() {
  return (
    <RecoilRoot>
      <QueryClientProvider client={queryClient}>
        <ThemeProvider theme={theme}>
          <Routes>
            <Route path="/" element={<ProblemListPage />} />
            <Route path="/clients" element={<ClientListPage />} />
            <Route
              path="/classes/:className/instances/:instance"
              element={<ProblemPage />}
            />
            <Route path="/clients/:clientName" element={<ClientPage />} />
            <Route path="/labels" element={<LabelListPage />} />
            <Route path="/upload" element={<UploadDataPage />} />
          </Routes>
        </ThemeProvider>
      </QueryClientProvider>
    </RecoilRoot>
  )
}

export default App
