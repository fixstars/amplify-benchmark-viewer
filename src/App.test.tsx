/**
 * Copyright (c) Fixstars Corporation and Fixstars Amplify Corporation.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { render } from '@testing-library/react'
import type { ReportData } from '@types'
import type { AxiosError } from 'axios'
import { BrowserRouter } from 'react-router-dom'
import mockData from 'utils/test/mocks/data/mockReportData.json'
import { mockQueryClientProvider } from 'utils/test/mocks/react-query'
import { theme } from 'utils/theme'

import App from './App'

let mockResponse: {
  readonly data: ReportData | undefined
  readonly isLoading: boolean
  readonly isFetching: boolean
  readonly isError: boolean
  readonly error: AxiosError | null
} = {
  data: undefined,
  isLoading: false,
  isFetching: false,
  isError: false,
  error: null,
}

jest.mock('api', () => ({
  useGetReportData: () => mockResponse,
}))

describe('<App />', () => {
  test('renders well with data', () => {
    mockResponse = {
      ...mockResponse,
      data: mockData,
    }

    const { container } = render(
      <BrowserRouter>
        <App />
      </BrowserRouter>,
    )

    const queryClient = mockQueryClientProvider.mock.calls[0][0]
    expect(queryClient.client.defaultOptions.queries.retry).toBe(false)

    const themeProvider = queryClient.children
    expect(themeProvider.type.name).toBe('ThemeProvider')
    expect(themeProvider.props.theme).toEqual(theme)

    const routes = themeProvider.props.children
    expect(routes.type.name).toBe('Routes')
    expect(routes.props.children.length).toBe(6)

    {
      const page = routes.props.children[0]
      expect(page.type.name).toBe('Route')
      expect(page.props.path).toBe('/')
      expect(page.props.element.type.name).toBe('ProblemListPage')
    }

    {
      const page = routes.props.children[1]
      expect(page.type.name).toBe('Route')
      expect(page.props.path).toBe('/clients')
      expect(page.props.element.type.name).toBe('ClientListPage')
    }

    {
      const page = routes.props.children[2]
      expect(page.type.name).toBe('Route')
      expect(page.props.path).toBe('/classes/:className/instances/:instance')
      expect(page.props.element.type.name).toBe('ProblemPage')
    }

    {
      const page = routes.props.children[3]
      expect(page.type.name).toBe('Route')
      expect(page.props.path).toBe('/clients/:clientName')
      expect(page.props.element.type.name).toBe('ClientPage')
    }

    {
      const page = routes.props.children[4]
      expect(page.type.name).toBe('Route')
      expect(page.props.path).toBe('/labels')
      expect(page.props.element.type.name).toBe('LabelListPage')
    }

    {
      const page = routes.props.children[5]
      expect(page.type.name).toBe('Route')
      expect(page.props.path).toBe('/upload')
      expect(page.props.element.type.name).toBe('UploadDataPage')
    }

    expect(container).toMatchSnapshot()
  })
})
