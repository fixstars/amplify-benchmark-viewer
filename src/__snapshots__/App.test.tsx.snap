// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders well with data 1`] = `
<div>
  <div
    class="MuiBox-root css-thz6x0"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-z80fak-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1r339ym-MuiToolbar-root"
      >
        <div
          class="css-yp9swi"
        >
          <a
            class="css-94l1l1"
            href="/"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-19sbe58-MuiTypography-root"
            >
              Annealing Benchmark
            </h6>
          </a>
        </div>
        <nav
          class="css-ov1ktg"
        >
          <a
            class="css-94l1l1"
            href="/"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-4hrd4f-MuiTypography-root"
            >
              Problems
            </p>
          </a>
          <a
            class="css-94l1l1"
            href="/clients"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-muv6ew-MuiTypography-root"
            >
              Clients
            </p>
          </a>
          <a
            class="css-94l1l1"
            href="/labels"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-muv6ew-MuiTypography-root"
            >
              Label
            </p>
          </a>
        </nav>
      </div>
    </header>
    <div
      class="MuiBox-root css-1cf7osc"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
      />
      <div
        class="MuiBox-root css-0"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-15h1z56-MuiGrid-root"
        >
          <h2
            class="MuiTypography-root MuiTypography-h2 css-1p54r94-MuiTypography-root"
          >
            Problem List
          </h2>
        </div>
      </div>
      <div
        class="MuiBox-root css-13pmxen"
      >
        <div
          aria-colcount="3"
          aria-multiselectable="false"
          aria-rowcount="6"
          class="MuiDataGrid-root MuiDataGrid-root--densityStandard MuiDataGrid-withBorderColor css-1in0i7x-MuiDataGrid-root"
          role="grid"
        >
          <div
            class="MuiDataGrid-main css-204u17-MuiDataGrid-main"
          >
            <div
              class="MuiDataGrid-columnHeaders MuiDataGrid-withBorderColor css-1iyq7zh-MuiDataGrid-columnHeaders"
              role="presentation"
              style="min-height: 56px; max-height: 56px; line-height: 56px;"
            >
              <div
                class="MuiDataGrid-columnHeadersInner MuiDataGrid-columnHeadersInner--scrollable css-gl260s-MuiDataGrid-columnHeadersInner"
                role="rowgroup"
                style="transform: translate3d(0px, 0px, 0px);"
              >
                <div
                  aria-rowindex="1"
                  class="css-yrdy0g-MuiDataGrid-columnHeaderRow"
                  role="row"
                >
                  <div
                    aria-colindex="1"
                    aria-label="Class"
                    aria-sort="none"
                    class="MuiDataGrid-columnHeader MuiDataGrid-withBorderColor"
                    data-field="class"
                    role="columnheader"
                    style="height: 56px; width: 180px; min-width: 180px; max-width: 180px;"
                    tabindex="0"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderDraggableContainer"
                      draggable="false"
                      role="presentation"
                    >
                      <div
                        class="MuiDataGrid-columnHeaderTitleContainer"
                        role="presentation"
                      >
                        <div
                          class="MuiDataGrid-columnHeaderTitleContainerContent"
                        >
                          <div
                            aria-label=""
                            class="MuiDataGrid-columnHeaderTitle css-t89xny-MuiDataGrid-columnHeaderTitle"
                            data-mui-internal-clone-element="true"
                          >
                            Class
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                      style="min-height: 56px; opacity: 1;"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                        data-testid="SeparatorIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M11 19V5h2v14z"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    aria-colindex="2"
                    aria-label="Instance"
                    aria-sort="none"
                    class="MuiDataGrid-columnHeader MuiDataGrid-withBorderColor"
                    data-field="instance"
                    role="columnheader"
                    style="height: 56px; width: 180px; min-width: 180px; max-width: 180px;"
                    tabindex="-1"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderDraggableContainer"
                      draggable="false"
                      role="presentation"
                    >
                      <div
                        class="MuiDataGrid-columnHeaderTitleContainer"
                        role="presentation"
                      >
                        <div
                          class="MuiDataGrid-columnHeaderTitleContainerContent"
                        >
                          <div
                            aria-label=""
                            class="MuiDataGrid-columnHeaderTitle css-t89xny-MuiDataGrid-columnHeaderTitle"
                            data-mui-internal-clone-element="true"
                          >
                            Instance
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                      style="min-height: 56px; opacity: 1;"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                        data-testid="SeparatorIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M11 19V5h2v14z"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    aria-colindex="3"
                    aria-label="Clients"
                    aria-sort="none"
                    class="MuiDataGrid-columnHeader MuiDataGrid-withBorderColor"
                    data-field="clients"
                    role="columnheader"
                    style="height: 56px; width: 0px; min-width: 0; max-width: 0;"
                    tabindex="-1"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderDraggableContainer"
                      draggable="false"
                      role="presentation"
                    >
                      <div
                        class="MuiDataGrid-columnHeaderTitleContainer"
                        role="presentation"
                      >
                        <div
                          class="MuiDataGrid-columnHeaderTitleContainerContent"
                        >
                          <div
                            aria-label=""
                            class="MuiDataGrid-columnHeaderTitle css-t89xny-MuiDataGrid-columnHeaderTitle"
                            data-mui-internal-clone-element="true"
                          >
                            Clients
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                      style="min-height: 56px; opacity: 1;"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                        data-testid="SeparatorIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M11 19V5h2v14z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="MuiDataGrid-virtualScroller css-axafay-MuiDataGrid-virtualScroller"
              role="presentation"
              style="overflow-x: hidden;"
            >
              <div
                class="MuiDataGrid-virtualScrollerContent MuiDataGrid-virtualScrollerContent--overflowed css-1kwdphh-MuiDataGrid-virtualScrollerContent"
                role="presentation"
                style="width: auto; height: 260px; min-height: auto;"
              >
                <div
                  class="MuiDataGrid-virtualScrollerRenderZone css-s1v7zr-MuiDataGrid-virtualScrollerRenderZone"
                  role="rowgroup"
                  style="transform: translate3d(0px, 0px, 0px);"
                >
                  <div
                    aria-rowindex="2"
                    aria-selected="false"
                    class="MuiDataGrid-row"
                    data-id="Tsp_a_instance"
                    data-rowindex="0"
                    role="row"
                    style="max-height: 52px; min-height: 52px;"
                  >
                    <div
                      aria-colindex="1"
                      aria-colspan="1"
                      class="MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="0"
                      data-field="class"
                      role="cell"
                      style="min-width: 180px; max-width: 180px; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <div
                        class="MuiDataGrid-cellContent"
                        role="presentation"
                        title="Tsp"
                      >
                        Tsp
                      </div>
                    </div>
                    <div
                      aria-colindex="2"
                      aria-colspan="1"
                      class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="1"
                      data-field="instance"
                      role="cell"
                      style="min-width: 180px; max-width: 180px; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <a
                        href="/classes/Tsp/instances/a_instance"
                      >
                        a_instance
                      </a>
                    </div>
                    <div
                      aria-colindex="3"
                      aria-colspan="1"
                      class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="2"
                      data-field="clients"
                      role="cell"
                      style="min-width: 0; max-width: 0; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <div
                        style="margin-right: 10px;"
                      >
                        <a
                          class="css-1o7gueb"
                          href="/clients/a_client"
                        >
                          a_client
                        </a>
                      </div>
                    </div>
                  </div>
                  <div
                    aria-rowindex="3"
                    aria-selected="false"
                    class="MuiDataGrid-row"
                    data-id="Tsp_b_instance"
                    data-rowindex="1"
                    role="row"
                    style="max-height: 52px; min-height: 52px;"
                  >
                    <div
                      aria-colindex="1"
                      aria-colspan="1"
                      class="MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="0"
                      data-field="class"
                      role="cell"
                      style="min-width: 180px; max-width: 180px; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <div
                        class="MuiDataGrid-cellContent"
                        role="presentation"
                        title="Tsp"
                      >
                        Tsp
                      </div>
                    </div>
                    <div
                      aria-colindex="2"
                      aria-colspan="1"
                      class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="1"
                      data-field="instance"
                      role="cell"
                      style="min-width: 180px; max-width: 180px; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <a
                        href="/classes/Tsp/instances/b_instance"
                      >
                        b_instance
                      </a>
                    </div>
                    <div
                      aria-colindex="3"
                      aria-colspan="1"
                      class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="2"
                      data-field="clients"
                      role="cell"
                      style="min-width: 0; max-width: 0; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <div
                        style="margin-right: 10px;"
                      >
                        <a
                          class="css-1o7gueb"
                          href="/clients/a_client"
                        >
                          a_client
                        </a>
                      </div>
                    </div>
                  </div>
                  <div
                    aria-rowindex="4"
                    aria-selected="false"
                    class="MuiDataGrid-row"
                    data-id="Tsp_h_instance"
                    data-rowindex="2"
                    role="row"
                    style="max-height: 52px; min-height: 52px;"
                  >
                    <div
                      aria-colindex="1"
                      aria-colspan="1"
                      class="MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="0"
                      data-field="class"
                      role="cell"
                      style="min-width: 180px; max-width: 180px; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <div
                        class="MuiDataGrid-cellContent"
                        role="presentation"
                        title="Tsp"
                      >
                        Tsp
                      </div>
                    </div>
                    <div
                      aria-colindex="2"
                      aria-colspan="1"
                      class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="1"
                      data-field="instance"
                      role="cell"
                      style="min-width: 180px; max-width: 180px; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <a
                        href="/classes/Tsp/instances/h_instance"
                      >
                        h_instance
                      </a>
                    </div>
                    <div
                      aria-colindex="3"
                      aria-colspan="1"
                      class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="2"
                      data-field="clients"
                      role="cell"
                      style="min-width: 0; max-width: 0; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <div
                        style="margin-right: 10px;"
                      >
                        <a
                          class="css-gqb4bp"
                          href="/clients/f_client"
                        >
                          f_client
                        </a>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="MuiDataGrid-footerContainer MuiDataGrid-withBorderColor css-wop1k0-MuiDataGrid-footerContainer"
          >
            <div />
            <div
              class="MuiTablePagination-root css-rtrcn9-MuiTablePagination-root"
            >
              <div
                class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-78c6dr-MuiToolbar-root-MuiTablePagination-toolbar"
              >
                <div
                  class="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
                />
                <p
                  class="MuiTablePagination-selectLabel css-pdct74-MuiTablePagination-selectLabel"
                  id=":r1:"
                >
                  Rows per page:
                </p>
                <div
                  class="MuiInputBase-root MuiInputBase-colorPrimary MuiTablePagination-input css-16c50h-MuiInputBase-root-MuiTablePagination-select"
                >
                  <div
                    aria-expanded="false"
                    aria-haspopup="listbox"
                    aria-labelledby=":r1: :r0:"
                    class="MuiSelect-select MuiTablePagination-select MuiSelect-standard MuiInputBase-input css-194a1fa-MuiSelect-select-MuiInputBase-input"
                    id=":r0:"
                    role="button"
                    tabindex="0"
                  >
                    100
                  </div>
                  <input
                    aria-hidden="true"
                    aria-invalid="false"
                    class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                    tabindex="-1"
                    value="100"
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiTablePagination-selectIcon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
                    data-testid="ArrowDropDownIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </div>
                <p
                  class="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
                >
                  1–5 of 5
                </p>
                <div
                  class="MuiTablePagination-actions"
                >
                  <button
                    aria-label="Go to previous page"
                    class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                    disabled=""
                    tabindex="-1"
                    title="Go to previous page"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="KeyboardArrowLeftIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                      />
                    </svg>
                  </button>
                  <button
                    aria-label="Go to next page"
                    class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                    disabled=""
                    tabindex="-1"
                    title="Go to next page"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="KeyboardArrowRightIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                      />
                    </svg>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
